Add Counterstrike mission Sarin Gas 2: Down Under

Smittytron opened the pull request:
Hoo boy this one took awhile.

I decided not to get fancy with difficulties here. There's normal mode and don't lose a single soldier mode.

Smittytron committed:
Add Counterstrike mission Sarin Gas 2: Down Under

Smittytron commented:
Rebased

Muh-Muh reviewed (commented):
I only found a few things that need changing and I had to go out of my way to actually make it crash ingame.

Other than that just two very minor points:

(1) Is it intentional that Tanya may day without any repercussions?
(2) The line on the ground in Tanya's prison doesn't correspond to the range of the flame towers. But I don't know if this is something that needs to be adressed at all.

Smittytron commented:
Updated. Added IsDead checks for everything but the mammoth 2 cam, which I moved over by a cell and expanded the trigger distance by 1. I'm certain that there's no way to destroy the mammoth without spawning the camera first.

> (1) Is it intentional that Tanya may day without any repercussions?
> (2) The line on the ground in Tanya's prison doesn't correspond to the range of the flame towers. But I don't know if this is something that needs to be adressed at all.

In the original you could lose Tanya and go on, but I guess it makes sense to have an objective for her. I added a free and keep Tanya alive objective.
As far as the line goes; I don't want to change a map from the original, and Tanya's vision is good enough for the player to see whether or not she is getting fire-balled or not. :smile:

pchote commented:
It would be great if we could get this merged in time for the playtest.

Muh-Muh commented:
> Updated. Added IsDead checks for everything 

One thing that still doesn't work right is the trigger for the scientist entering the demo truck.
```
if not ScientistMan.IsDead then
		Trigger.OnRemovedFromWorld(ScientistMan, DemoTruckExit)
```

This creates the trigger and if the scientist then dies (while on the way to the truck) the truck starts moving. Expected behavior is: Truck doesn't move if the scientist is killed on the way. Perhaps, instead of "OnRemovedFromWorld" you could use "OnPassengerEntered" or a simple proximitytrigger?

By the way: did you add an extra dog? I can't free tanya with just the spies from the prison anymore. All the lua-crashed depended on being able to do this somewhat easily. If this is taken care of I think the mission works fine as is.

Smittytron commented:
Updated: Moving the IsDead check to the next function worked for me.

> By the way: did you add an extra dog?

Yeah, I thought it made more sense than having two flamers there. In theory, you aren't supposed to be able to use the two spies without breaking them out of prison first. I didn't realize they could get past their own guard dogs.

Muh-Muh commented:
Found something else (sorry -.-):  
You can infiltrate the first war factory multiple times. This gives a lua crash for the second infiltration.
  Can infiltratable be made conditional? If so, you could make it conditional and remove the condition for both warfactories after the first successful infiltration.

Smittytron commented:
Updated

Muh-Muh commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+]]
+SarinDispensed = false
+TanyaFreed = false
+TruckSolen = false
```
There is a typo here. The trigger never fires.

Muh-Muh commented:
I have no more complaints. I think it works well now.

Smittytron commented:
Updated. Changed the hard objective message to specify that the player only needs to protect the units under their command.

Smittytron commented:
Updated to work with #16041 and #16119.

pchote commented:
Can we get some testing and reviewing here?

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+--[[
+   Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
```
2019

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+			if Map.LobbyOption("difficulty") == "hard" then
+				greece.MarkCompletedObjective(NoCasualties)
+			end
+			if TanyaFreed == false then
```
if not TanyaFreed then

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+	end)
+
+	Trigger.OnInfiltrated(WarFactory2, function()
+		if not StealMammoth.IsDead or StealMammoth.Owner ~= ussr then
```
Doesn't this want to be `not StealMammoth.IsDead and StealMammoth.Owner == ussr`?

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+end
+
+ConsoleTriggers = function()
+	Trigger.OnEnteredProximityTrigger(Terminal1.CenterPosition, WDist.FromCells(1), function(actor, trigger1)
```
Just a rand-note: No need to name this `trigger1`, `trigger2`, ... as those are local variables which only exist inside this function (scope).

Smittytron replied:
I did this as a readability thing for myself.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+	Trigger.OnEnteredProximityTrigger(Terminal1.CenterPosition, WDist.FromCells(1), function(actor, trigger1)
+		if actor.Owner == greece then
+			Trigger.RemoveProximityTrigger(trigger1)
+			Media.DisplayMessage("Flame Turret deactivated", "Console")
```
Don't we want to play this message only when we actually deactivate, i.e. kill, the flame tower?

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+	Trigger.OnEnteredProximityTrigger(Terminal2.CenterPosition, WDist.FromCells(1), function(actor, trigger2)
+		if actor.Owner == greece then
+			Trigger.RemoveProximityTrigger(trigger2)
+			Media.DisplayMessage("Flame Turret deactivated", "Console")
```
Same here and for the third.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+
+	Trigger.OnEnteredProximityTrigger(Terminal4.CenterPosition, WDist.FromCells(1), function(actor, trigger4)
+		if actor.Owner == greece then
+			SarinDispensed = true
```
This variable isn't used.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+			flare4 = Actor.Create("flare", true, { Owner = england, Location = Sarin4.Location })
+			Trigger.AfterDelay(DateTime.Seconds(4), function()
+				Utils.Do(SarinVictims, function(actor)
+					if actor.IsDead then
```
This can be shortened to
```lua
if not actor.IsDead then
	actor.Kill()
end
```

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+			Trigger.RemoveProximityTrigger(trigger4)
+			Media.DisplayMessage("Sarin Nerve Gas dispensers activated", "Console")
+			local KillCamera = Actor.Create("camera", true, { Owner = greece, Location = Sarin2.Location })
+			flare1 = Actor.Create("flare", true, { Owner = england, Location = Sarin1.Location })
```
This and the others below probably want to use the `local` keyword.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+	Trigger.OnEnteredProximityTrigger(Terminal5.CenterPosition, WDist.FromCells(1), function(actor, trigger5)
+		if actor.Owner == greece then
+			Trigger.RemoveProximityTrigger(trigger5)
+			Media.DisplayMessage("Tesla Coil deactivated", "Console")
```
Here again.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+		if actor.Owner == greece then
+			Trigger.RemoveProximityTrigger(trigger6)
+			Media.DisplayMessage("Initialising Tesla Coil defence", "Console")
+			local Tesla1 = Actor.Create("tsla", true, { Owner = turkey, Location = TurkeyCoil1.Location })
```
Local variable names want to start with a lowercase letter.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+
+CameraTriggers = function()
+	Trigger.AfterDelay(DateTime.Seconds(1), function()
+		local StartCamera = Actor.Create("camera", true, { Owner = greece, Location = start.Location })
```
Same here (and all the `Camera`s below).

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+
+TruckSteal = function()
+	Trigger.OnInfiltrated(WarFactory1, function()
+		if TruckStolen == false and not StealTruck.IsDead then
```
if not TruckStolen and not StealTruck.IsDead then

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+	Trigger.OnInfiltrated(WarFactory1, function()
+		if TruckStolen == false and not StealTruck.IsDead then
+			TruckStolen = true
+			local TruckSteal1 = Actor.Create("camera", true, { Owner = greece, Location = TruckDrive1.Location })
```
lower case please.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+
+SpyTruckDrive = function()
+	StealTruck = Reinforcements.Reinforce(ussr, TruckType, TruckEntry)
+	local TruckSteal2 = Actor.Create("camera", true, { Owner = greece, Location = TruckCamera.Location })
```
And here.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+
+	Trigger.OnEnteredFootprint({ TruckDrive5.Location }, function(actor, triggertruck2)
+		if actor.Type == "truk" then
+			local DogCrewCamera = Actor.Create("camera", true, { Owner = greece, Location = DoggyCam.Location })
```
And here.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+				DogCrewCamera.Destroy()
+			end)
+			Utils.Do(DogCrew, function(actor)
+				if actor.IsDead then
```
This again can be simplified (see above for how).

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+		Media.DisplayMessage("Warning, prisoners are attempting to escape!", "Intercom")
+		Media.PlaySoundNotification(greece, "AlertBuzzer")
+		Utils.Do(GuardDogs, function(actor)
+			if actor.IsDead then
```
Same here.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+			if actor.IsDead then
+				return
+			else
+				actor.Hunt()
```
Should they only hunt once, or always when idle?

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+		end
+
+		Utils.Do(FiringSquad, function(actor)
+			if actor.IsDead then
```
Same here.

abcdefg30 commented on mods/ra/maps/sarin-gas-2-down-under/downunder.lua:
```diff
@@ -0,0 +1,457 @@
+	england = Player.GetPlayer("England")
+	turkey = Player.GetPlayer("Turkey")
+	ussr = Player.GetPlayer("USSR")
+	badguy = Player.GetPlayer("BadGuy")
```
This doesn't seem to be used. (In general I would try to remove as many [unnecessary] players from map.yaml as well, e.g. I guess all of badguy's units can be owned by ussr and the badguy player removed.)

Smittytron replied:
The reason for having the badguy player is so the Turkish teslas will only target the badguy mammoth and not a spy with a USSR disguise. I changed all the badguy actors sans said mammoth to USSR.

Smittytron commented:
Updated

abcdefg30 reviewed (dismissed):
Looks good to me otherwise.

Smittytron commented:
Updated

obrakmann reviewed (commented):
I have only one complaint about this mission (other than my general grievances about these kind of maps having no place in an RTS): when all your dudes die before Tanya gets freed, the mission does not fail automatically. It would be nice if that could be fixed. Otherwise the mission looks ok.

Smittytron commented:
Updated. I addressed the above comment by making Tanya temporarily German. She will fall under the player's command when the flametowers guarding her are destroyed.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/045a3375f0bfdad6d58740fa4fb93d24d731cdb2)
